core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java

```diff
@@ -238,12 +238,11 @@ private synchronized void init(MasterInquireClient masterInquireClient) {
           .setMasterInquireClient(mMasterInquireClient)
           .build());
       mClientMasterSync = new ClientMasterSync(mMetricsMasterClient, mAppId);
-      mExecutorService = Executors.newFixedThreadPool(1,
-          ThreadFactoryUtils.build("metrics-master-heartbeat-%d", true));
-      mExecutorService
-          .submit(new HeartbeatThread(HeartbeatContext.MASTER_METRICS_SYNC, mClientMasterSync,
-              (int) mClientContext.getConf().getMs(PropertyKey.USER_METRICS_HEARTBEAT_INTERVAL_MS),
-              mClientContext.getConf()));
+
+      mMetricsHeartbeat = EXECUTOR_SERVICE.scheduleWithFixedDelay(mClientMasterSync::heartbeat, 0,
```

Comment:
Do we need a separate heartbeat per-context, or could they be aggregated so that there's only ever 1 heartbeat?

Reply:
It's possible we could aggregate them, but will likely require some more changes to the RPC. We would also not be able to ever `stop` this thread from running after the JVM starts

Reply:
If we have 50 of these heartbeat threads, all 50 are sending the same exact set of metrics every 3 seconds. It doesn't really matter whether it's being done from 1 thread or 50 threads, either way it's wasteful and putting unnecessary load on the master

Reply:
If we count the number of open contexts, we can stop the heartbeat thread when the last context closes

Reply:
Well it's not that they are sending the same set of metrics. Even if the object is static, they technically will report different app ID's with each metric heartbeat. If we want to continue to support application client-level granularity I'd need to modify the RPC.

On top of this: another issue I've found is that `MetricsSystem` is a static class and is only instantiated on a per-JVM basis. This leads to an issue where multiple clients (`FileSystemContexts`) in the same JVM  will end up reporting metrics for other clients in the same JVM. Until we remove the singleton `MetricsSystem` the functionality for client-object level granularity will be broken.

Reply:
I've completely overhauled the implementation for adding heartbeats. Please see 0edab3c